File: tests/java/test-app/src/main/java/com/godaddy/asherah/testapp/regression/AppEncryptionJsonTest.java
package com.godaddy.asherah.testapp.regression;

import com.godaddy.asherah.appencryption.AppEncryption;
import com.godaddy.asherah.appencryption.AppEncryptionSessionFactory;
import com.godaddy.asherah.appencryption.persistence.Persistence;
import com.godaddy.asherah.testapp.utils.DateTimeUtils;
import com.godaddy.asherah.testapp.utils.PayloadGenerator;
import com.godaddy.asherah.testapp.utils.PersistenceFactory;
import com.godaddy.asherah.testapp.utils.SessionFactoryGenerator;

import org.json.JSONObject;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static com.godaddy.asherah.testapp.testhelpers.Constants.*;
import static org.junit.jupiter.api.Assertions.*;

public class AppEncryptionJsonTest {
  private static Persistence<byte[]> persistenceBytes;

  private JSONObject payload;
  private AppEncryptionSessionFactory appEncryptionSessionFactory;
  private String partitionId;
  private AppEncryption<JSONObject, byte[]> appEncryptionJson;
  // TODO Consider adding JSONObject-style envelope/persistence tests as well

  @BeforeAll
  public static void init() {
    persistenceBytes = PersistenceFactory.<byte[]>createInMemoryPersistence();
  }

  @BeforeEach
  public void setupTest() {
    payload = PayloadGenerator.createDefaultRandomJsonPayload();
    appEncryptionSessionFactory = SessionFactoryGenerator.createDefaultAppEncryptionSessionFactory();
    partitionId = DEFAULT_PARTITION_ID + "_" + DateTimeUtils.getCurrentTimeAsUtcIsoOffsetDateTime();
    appEncryptionJson = appEncryptionSessionFactory.getAppEncryptionJson(partitionId);
  }

  @AfterEach
  public void tearDown() {
    appEncryptionJson.close();
    appEncryptionSessionFactory.close();
  }

  @Test
  public void jsonEncryptDecrypt() {
    byte[] dataRowRecord = appEncryptionJson.encrypt(payload);
    JSONObject decryptedPayload = appEncryptionJson.decrypt(dataRowRecord);

    assertTrue(payload.similar(decryptedPayload));
  }

  @Test
  public void jsonEncryptDecryptSameSessionMultipleRounds() {
    // Just loop a bunch of times to verify no surprises
    final int iterations = 40;
    for (int i = 0; i < iterations; i++) {
      byte[] dataRowRecord = appEncryptionJson.encrypt(payload);
      JSONObject decryptedPayload = appEncryptionJson.decrypt(dataRowRecord);

      assertTrue(payload.similar(decryptedPayload));
    }
  }

  @Test
  public void jsonStoreLoad() {
    String persistenceKey = appEncryptionJson.store(payload, persistenceBytes);

    Optional<JSONObject> decryptedJsonPayload = appEncryptionJson.load(persistenceKey, persistenceBytes);

    if (decryptedJsonPayload.isPresent()) {
      assertTrue(payload.similar(decryptedJsonPayload.get()));
    }
    else {
      fail("Json load did not return decrypted payload");
    }
  }

  @Test
  public void jsonLoadInvalidKey() {
    String persistenceKey = "1234";

    Optional<JSONObject> decryptedJsonPayload = appEncryptionJson.load(persistenceKey, persistenceBytes);

    assertFalse(decryptedJsonPayload.isPresent());
  }

  @Test
  public void jsonEncryptDecryptWithDifferentSession() {
    byte[] dataRowRecord = appEncryptionJson.encrypt(payload);

    try (AppEncryption<JSONObject, byte[]> appEncryptionJsonNew = appEncryptionSessionFactory.getAppEncryptionJson(partitionId)) {
      JSONObject decryptedPayload = appEncryptionJsonNew.decrypt(dataRowRecord);

      assertTrue(payload.similar(decryptedPayload));
    }
  }

  @Test
  public void jsonEncryptDecryptWithDifferentPayloads() {
    final JSONObject otherPayload = PayloadGenerator.createDefaultRandomJsonPayload();
    byte[] dataRowRecord1 = appEncryptionJson.encrypt(payload);
    byte[] dataRowRecord2 = appEncryptionJson.encrypt(otherPayload);

    JSONObject decryptedPayload1 = appEncryptionJson.decrypt(dataRowRecord1);
    JSONObject decryptedPayload2 = appEncryptionJson.decrypt(dataRowRecord2);

    assertTrue(payload.similar(decryptedPayload1));
    assertTrue(otherPayload.similar(decryptedPayload2));
  }

  @Test
  public void jsonStoreOverwritePayload() {
    String key = "some_key";
    final JSONObject otherPayload = PayloadGenerator.createDefaultRandomJsonPayload();

    appEncryptionJson.store(key, payload, persistenceBytes);
    appEncryptionJson.store(key, otherPayload, persistenceBytes);
    JSONObject actualResult = appEncryptionJson.load(key, persistenceBytes).get();

    assertTrue(otherPayload.similar(actualResult));
  }

}

File: languages/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/AppEncryptionJsonTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.IntegrationTests.Utils;
using GoDaddy.Asherah.AppEncryption.Persistence;
using LanguageExt;
using Newtonsoft.Json.Linq;
using Xunit;
using Xunit.Sdk;

using static GoDaddy.Asherah.AppEncryption.IntegrationTests.TestHelpers.Constants;

namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.Regression
{
    [Collection("Configuration collection")]
    public class AppEncryptionJsonTest : IDisposable
    {
        private static readonly Persistence<byte[]> PersistenceBytes =
            PersistenceFactory<byte[]>.CreateInMemoryPersistence();

        private readonly JObject payload;
        private readonly AppEncryptionSessionFactory appEncryptionSessionFactory;
        private readonly string partitionId;
        private readonly AppEncryption<JObject, byte[]> appEncryptionJson;

        public AppEncryptionJsonTest(ConfigFixture configFixture)
        {
            payload = PayloadGenerator.CreateDefaultRandomJsonPayload();
            appEncryptionSessionFactory = SessionFactoryGenerator.CreateDefaultAppEncryptionSessionFactory(configFixture.KeyManagementService, configFixture.MetastorePersistence);
            partitionId = DefaultPartitionId + "_" + DateTimeUtils.GetCurrentTimeAsUtcIsoDateTimeOffset();
            appEncryptionJson = appEncryptionSessionFactory.GetAppEncryptionJson(partitionId);
        }

        public void Dispose()
        {
            appEncryptionJson.Dispose();
            appEncryptionSessionFactory.Dispose();
        }

        [Fact]
        private void JsonEncryptDecrypt()
        {
            byte[] dataRowRecord = appEncryptionJson.Encrypt(payload);
            JObject decryptedPayload = appEncryptionJson.Decrypt(dataRowRecord);

            Assert.Equal(payload, decryptedPayload);
        }

        [Fact]
        private void JsonEncryptDecryptSameSessionMultipleRounds()
        {
            // Just loop a bunch of times to verify no surprises
            int iterations = 40;
            for (int i = 0; i < iterations; i++)
            {
                byte[] dataRowRecord = appEncryptionJson.Encrypt(payload);
                JObject decryptedPayload = appEncryptionJson.Decrypt(dataRowRecord);

                Assert.Equal(payload, decryptedPayload);
            }
        }

        [Fact]
        private void JsonStoreLoad()
        {
            string persistenceKey = appEncryptionJson.Store(payload, PersistenceBytes);

            Option<JObject> decryptedPayload = appEncryptionJson.Load(persistenceKey, PersistenceBytes);

            if (decryptedPayload.IsSome)
            {
                Assert.Equal(payload, (JObject)decryptedPayload);
            }
            else
            {
                throw new XunitException("Json load did not return decrypted payload");
            }
        }

        [Fact]
        private void JsonLoadInvalidKey()
        {
            string persistenceKey = "1234";

            Option<JObject> decryptedPayload = appEncryptionJson.Load(persistenceKey, PersistenceBytes);

            Assert.False(decryptedPayload.IsSome);
        }

        [Fact]
        private void JsonEncryptDecryptWithDifferentSession()
        {
            byte[] dataRowRecord = appEncryptionJson.Encrypt(payload);

            using (AppEncryption<JObject, byte[]> appEncryptionBytesNew =
                appEncryptionSessionFactory.GetAppEncryptionJson(partitionId))
            {
                JObject decryptedPayload = appEncryptionBytesNew.Decrypt(dataRowRecord);
                Assert.Equal(payload, decryptedPayload);
            }
        }

        [Fact]
        private void JsonEncryptDecryptWithDifferentPayloads()
        {
            JObject otherPayload = PayloadGenerator.CreateDefaultRandomJsonPayload();
            byte[] dataRowRecord1 = appEncryptionJson.Encrypt(payload);
            byte[] dataRowRecord2 = appEncryptionJson.Encrypt(otherPayload);

            JObject decryptedPayload1 = appEncryptionJson.Decrypt(dataRowRecord1);
            JObject decryptedPayload2 = appEncryptionJson.Decrypt(dataRowRecord2);

            Assert.Equal(payload, decryptedPayload1);
            Assert.Equal(otherPayload, decryptedPayload2);
        }

        [Fact]
        private void JsonStoreOverwritePayload()
        {
            string key = "some_key";
            JObject otherPayload = PayloadGenerator.CreateDefaultRandomJsonPayload();

            appEncryptionJson.Store(key, payload, PersistenceBytes);
            appEncryptionJson.Store(key, otherPayload, PersistenceBytes);
            Option<JObject> decryptedPayload = appEncryptionJson.Load(key, PersistenceBytes);

            Assert.Equal(otherPayload, (JObject)decryptedPayload);
        }
    }
}
